sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExpressionSet.scala

```diff
@@ -50,7 +50,7 @@ object ExpressionSet {
 class ExpressionSet protected(
     protected val baseSet: mutable.Set[Expression] = new mutable.HashSet,
     protected val originals: mutable.Buffer[Expression] = new ArrayBuffer)
-  extends Set[Expression] {
+  extends Set[Expression] with Serializable {
```

Comment:
Yes,  `constraints` in `QueryPlan` is `ExpressionSet` and `SparkPlan` which is a subclass of `QueryPlan` is serializable so `ExpressionSet` should be also serializable strictly. But `constraints` is lazy val and it's not accessed when the receiver object is a instance of `SparkPlan`. In other word, `constraints` is accessed only when the receiver object is a instance of `LogicalPlan`.